src/Common/NamedCollections/NamedCollectionUtils.cpp

```diff
@@ -322,22 +353,155 @@ class LoadFromSQL : private WithContext
+        writeString(data, out);
 
         out.next();
-        if (settings.fsync_metadata)
+        if (getContext()->getSettingsRef().fsync_metadata)
             out.sync();
         out.close();
 
-        fs::rename(tmp_path, path);
+        fs::rename(tmp_path, getPath(path));
+    }
+
+    void remove(const std::string & path) override
+    {
+        if (!removeIfExists(getPath(path)))
+        {
+            throw Exception(
+                ErrorCodes::NAMED_COLLECTION_DOESNT_EXIST,
+                "Cannot remove `{}`, because it doesn't exist", path);
+        }
+    }
+
+    bool removeIfExists(const std::string & path) override
+    {
+        return fs::remove(getPath(path));
+    }
+
+private:
+    std::string getPath(const std::string & path) const
+    {
+        return fs::path(root_path) / path;
+    }
+
+    /// Delete .tmp files. They could be left undeleted in case of
+    /// some exception or abrupt server restart.
+    void cleanup()
+    {
+        std::vector<std::string> files_to_remove;
+        for (fs::directory_iterator it{root_path}; it != fs::directory_iterator{}; ++it)
+        {
+            const auto & current_path = it->path();
+            if (current_path.extension() == ".tmp")
+                files_to_remove.push_back(current_path);
+        }
+        for (const auto & file : files_to_remove)
+            fs::remove(file);
     }
 };
 
+
+class NamedCollectionsZooKeeperStorage : public INamedCollectionsStorage, private WithContext
+{
+private:
+    std::string root_path;
+    mutable zkutil::ZooKeeperPtr zookeeper_client{nullptr};
+
+public:
+    NamedCollectionsZooKeeperStorage(ContextPtr context_, const std::string & path_)
+        : WithContext(context_)
+        , root_path(path_)
+    {
+        if (root_path.empty())
+            throw Exception(ErrorCodes::INVALID_CONFIG_PARAMETER, "Collections path cannot be empty");
+
+        if (root_path != "/" && root_path.back() == '/')
+            root_path.resize(root_path.size() - 1);
+        if (root_path.front() != '/')
+            root_path = "/" + root_path;
+
+        auto client = getClient();
+        if (root_path != "/" && !client->exists(root_path))
+        {
+            client->createAncestors(root_path);
+            client->createIfNotExists(root_path, "");
+        }
+    }
+
+    ~NamedCollectionsZooKeeperStorage() override = default;
+
+    std::vector<std::string> list() const override
+    {
+        return getClient()->getChildren(root_path);
+    }
+
+    bool exists(const std::string & path) const override
+    {
+        return getClient()->exists(getPath(path));
+    }
+
+    std::string read(const std::string & path) const override
+    {
+        return getClient()->get(getPath(path));
+    }
+
+    void write(const std::string & path, const std::string & data, bool replace) override
+    {
+        if (replace)
+        {
+            getClient()->createOrUpdate(getPath(path), data, zkutil::CreateMode::Persistent);
```

Comment:
Maybe `return` here and move the other branch one indentation lower?

Reply:
In my opinion the scope of the indentation is too small to make it matter to remove it, makes sense if the scope is big, otherwise ok either way.